### PR TITLE
Expose control of what part of text (start, middle or end) of line labels corresponds to the anchor position

### PR DESCRIPTION
**Pull request by nyalldawson**
Previously labels would always be centered over the anchor position of lines (unless it was a curved label, in which case it was the start of the lable!) This is limiting (and inconsistent) for some placements, where it is preferable to e.g. line up the end
of the label text with a specific point on the line.

There's also an option for "Follow placement", which means that the text anchor will depends on the placement along line setting:
- For labels anchored near the start of the line (0-25%), the anchor placement will be the **start** of the label text
- For labels anchored near the end of the line (75-100%), the anchor placement will be the **end** of the label text
- For anchor placements between 25 and 75%, the anchor placement will be the **center** of the label text

### CODE REVIEW

File: src/core/pal/feature.cpp [hunk 1]
```diff
@@ -1173,7 +1211,20 @@ std::size_t FeaturePart::createCandidatesAlongLineNearMidpoint( std::vector< std
...
     // penalize positions which are further from the line's anchor point
-    double costCenter = std::fabs( lineAnchorPoint - ( currentDistanceAlongLine + labelWidth / 2 ) ) / totalLineLength; // <0, 0.5>
+    double textAnchorPoint = 0;
+    switch ( mLF->lineAnchorTextPoint() )
+    {
+      case QgsLabelLineSettings::AnchorTextPoint::StartOfText:
+        textAnchorPoint = currentDistanceAlongLine;
+        break;
+      case QgsLabelLineSettings::AnchorTextPoint::CenterOfText:
+        textAnchorPoint = ( currentDistanceAlongLine + labelWidth / 2 );
```
Comment (elpaso) [inline src/core/pal/feature.cpp:1221]:
```suggestion
        textAnchorPoint = currentDistanceAlongLine + labelWidth / 2;
```

**Review by elpaso — COMMENTED** [review]
Nice work! 

Are you planning any tests?

File: src/core/labeling/qgslabelfeature.h [hunk 2]
```diff
@@ -471,6 +470,26 @@ class CORE_EXPORT QgsLabelFeature
      */
     void setLineAnchorType( QgsLabelLineSettings::AnchorType type ) { mLineAnchorType = type; }
 
+    /**
+     * Returns the line anchor text point, which dictates which part of the label text
```
Comment (DelazJ) [inline src/core/labeling/qgslabelfeature.h:474]: Does it need to be as this, mimicking the next function?
```suggestion
     * Returns the line anchor text \a point, which dictates which part of the label text
```

File: src/core/labeling/qgspallabeling.h [hunk 3]
```diff
@@ -336,6 +336,7 @@ class CORE_EXPORT QgsPalLayerSettings
       PolygonLabelOutside = 109, //!< Whether labels outside a polygon feature are permitted, or should be forced (since QGIS 3.14)
       LineAnchorPercent = 111, //!< Portion along line at which labels should be anchored (since QGIS 3.16)
       LineAnchorClipping = 112, //!< Clipping mode for line anchor calculation (since QGIS 3.20)
+      LineAnchorTextPoint = 115, //!< Line anchor text point (since QGIS 3.24)
```
Comment (DelazJ) [inline src/core/labeling/qgspallabeling.h:339]:
```suggestion
      LineAnchorTextPoint = 115, //!< Line anchor text point (since QGIS 3.26)
```
?

**Review by nirvn — APPROVED** [review]
Looks good.

**Pull request merged**